File: src/test/java/com/sparetimedevs/questlog/login/validator/EmailAddressPasswordMatchValidatorTest.java
package com.sparetimedevs.questlog.login.validator;

import com.sparetimedevs.questlog.login.Login;
import com.sparetimedevs.questlog.user.UserService;
import com.sparetimedevs.questlog.userpassword.UserPassword;
import com.sparetimedevs.questlog.userpassword.UserPasswordService;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import testsupport.MockitoExtension;

import java.util.UUID;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.core.Is.is;
import static org.hamcrest.core.IsNot.not;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class EmailAddressPasswordMatchValidatorTest {

	@InjectMocks
	private LoginValidator loginValidator;

	@Mock
	private UserService userService;

	@Mock
	private UserPasswordService userPasswordService;

	@Test
	void givenMatchingEmailAddressAndPasswordWhenValidateIsCalledResultsInNoException() throws Exception {
		UUID userId = UUID.randomUUID();
		UUID userPasswordId = UUID.randomUUID();
		String emailAddress = "ab@cd.ef";
		String password = "ghijkl";
		UserPassword userPassword = new UserPassword(userPasswordId, userId, password);
		Login login = new Login(emailAddress, password);

		when(userService.getUserId(login)).thenReturn(userId);
		when(userPasswordService.getUserPassword(userId)).thenReturn(userPassword);

		UUID returnedUserId = loginValidator.validate(login);

		assertThat(userPassword.getPassword(), is(equalTo(password)));
		assertThat(returnedUserId, is(equalTo(userId)));
	}

	@Test
	void givenMatchingEmailAddressAndPasswordWhenValidateIsCalledThrowsEmailAddressPasswordDoNotMatchException() throws Exception {
		UUID userId = UUID.randomUUID();
		UUID userPasswordId = UUID.randomUUID();
		String emailAddress = "ab@cd.ef";
		String password = "ghijkl";
		String notMatchingPassword = "mnopqrs";
		UserPassword userPassword = new UserPassword(userPasswordId, userId, password);
		Login login = new Login(emailAddress, notMatchingPassword);

		when(userService.getUserId(login)).thenReturn(userId);
		when(userPasswordService.getUserPassword(userId)).thenReturn(userPassword);

		assertThrows(EmailAddressPasswordDoNotMatchException.class, () -> {
			loginValidator.validate(login);
		});
		assertThat(userPassword.getPassword(), is(not(equalTo(notMatchingPassword))));
	}
}
//TODO write kotlin test?

File: src/test/java/com/sparetimedevs/questlog/login/validator/EmailAddressPasswordMatchValidatorTest.kt
package com.sparetimedevs.questlog.login.validator

import com.sparetimedevs.questlog.login.Login
import com.sparetimedevs.questlog.login.exception.EmailAddressPasswordDoNotMatchException
import com.sparetimedevs.questlog.user.UserService
import com.sparetimedevs.questlog.userpassword.UserPassword
import com.sparetimedevs.questlog.userpassword.UserPasswordService
import io.kotlintest.shouldBe
import io.kotlintest.specs.StringSpec
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.core.Is.`is`
import org.hamcrest.core.IsNot.not
import org.junit.jupiter.api.Assertions.assertThrows
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import test.EMAIL_ADDRESS_1
import test.PASSWORD_1
import test.PASSWORD_2
import test.userId1
import test.userPasswordId1

class EmailAddressPasswordMatchValidatorTest : StringSpec({

	val userService: UserService = mock(UserService::class.java)
	val userPasswordService: UserPasswordService = mock(UserPasswordService::class.java)
	val loginValidator = LoginValidator(userService, userPasswordService)

	val userPassword = UserPassword(userPasswordId1, userId1, PASSWORD_1)

	"given matching email address and password when validate then users id is returned" {
		val login = Login(EMAIL_ADDRESS_1, PASSWORD_1)

		`when`(userService.getUserId(login)).thenReturn(userId1)
		`when`(userPasswordService.getUserPassword(userId1)).thenReturn(userPassword)

		val returnedUserId = loginValidator.validate(login)

		userPassword.password shouldBe PASSWORD_1
		returnedUserId shouldBe userId1
	}

	"given matching email address and password when validate then EmailAddressPasswordDoNotMatchException is thrown" {
		val login = Login(EMAIL_ADDRESS_1, PASSWORD_2)

		`when`(userService.getUserId(login)).thenReturn(userId1)
		`when`(userPasswordService.getUserPassword(userId1)).thenReturn(userPassword)

		assertThrows(EmailAddressPasswordDoNotMatchException::class.java) { loginValidator.validate(login) }
		assertThat(userPassword.password, `is`(not(equalTo(PASSWORD_2))))
		//TODO use kotlintest instead of assertThat..
	}

})
